lib/streamlit/elements/utils.py

```diff
@@ -96,3 +112,105 @@ def get_label_visibility_proto_value(
+    options: Type[Enum],
+    opt_sequence: Sequence[Any],
+) -> RegisterWidgetResult[Enum]:
+    ...
+
+
+@overload
+def maybe_coerce_enum(
+    register_widget_result: RegisterWidgetResult[T],
+    options: type_util.OptionSequence[T],
+    opt_sequence: Sequence[T],
+) -> RegisterWidgetResult[T]:
+    ...
+
+
+def maybe_coerce_enum(register_widget_result, options, opt_sequence):
+    """Maybe Coerce a RegisterWidgetResult with an Enum member value to
+    RegisterWidgetResult[option] if option is an EnumType, otherwise just return
+    the original RegisterWidgetResult."""
+
+    # If the value is not a Enum, return early
+    if not isinstance(register_widget_result.value, Enum):
+        return register_widget_result
+
+    coerce_class: Optional[EnumMeta]
+    if isinstance(options, EnumMeta):
+        coerce_class = options
+    else:
+        coerce_class = _extract_common_class_from_iter(opt_sequence)
+        if not isinstance(coerce_class, EnumMeta):
+            return register_widget_result
+
+    return RegisterWidgetResult(
+        type_util.coerce_enum(register_widget_result.value, coerce_class),
+        register_widget_result.value_changed,
+    )
+
+
+# slightly ugly typing because TypeVars with Generic Bounds are not supported
+# (https://github.com/python/typing/issues/548)
+@overload
+def maybe_coerce_enum_sequence(
+    register_widget_result: RegisterWidgetResult[List[T]],
+    options: type_util.OptionSequence[T],
+    opt_sequence: Sequence[T],
+) -> RegisterWidgetResult[List[T]]:
+    ...
+
+
+@overload
+def maybe_coerce_enum_sequence(
+    register_widget_result: RegisterWidgetResult[Tuple[T, T]],
+    options: type_util.OptionSequence[T],
+    opt_sequence: Sequence[T],
+) -> RegisterWidgetResult[Tuple[T, T]]:
+    ...
+
+
+def maybe_coerce_enum_sequence(register_widget_result, options, opt_sequence):
+    """Maybe Coerce a RegisterWidgetResult with a sequence of Enum members as value
+    to RegisterWidgetResult[Sequence[option]] if option is an EnumType, otherwise just return
+    the original RegisterWidgetResult."""
+
+    # If not all widget values are Enums, return early
+    if not all(isinstance(val, Enum) for val in register_widget_result.value):
+        return register_widget_result
+
+    # Extract the class to coerce
+    coerce_class: Optional[EnumMeta]
+    if isinstance(options, EnumMeta):
+        coerce_class = options
+    else:
+        coerce_class = _extract_common_class_from_iter(opt_sequence)
+        if not isinstance(coerce_class, EnumMeta):
+            return register_widget_result
+
+    # Return a new RegisterWidgetResult with the coerced enum values sequence
+    return RegisterWidgetResult(
+        type(register_widget_result.value)(
+            type_util.coerce_enum(val, coerce_class)
+            for val in register_widget_result.value
+        ),
+        register_widget_result.value_changed,
+    )
+
+
+def _extract_common_class_from_iter(iterable: Iterable[Any]) -> Any:
```

Comment:
The design of this function feels weird to me, specifically returning the iterable if there is no shared class. Why not just return `None` or some other singleton value to indicate this situation?